engine/execution/checker/engine.go

```diff
@@ -0,0 +1,100 @@
+package checker
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/consensus/hotstuff/notifications"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+type Engine struct {
+	notifications.NoopConsumer // satisfy the FinalizationConsumer interface
+
+	unit      *engine.Unit
+	log       zerolog.Logger
+	state     protocol.State
+	execState state.ExecutionState
+	seals     storage.Seals // to compare the executed result with the sealed result
+}
+
+func New(
+	logger zerolog.Logger,
+	state protocol.State,
+	execState state.ExecutionState,
+	seals storage.Seals,
+) *Engine {
+	return &Engine{
+		unit:      engine.NewUnit(),
+		log:       logger.With().Str("engine", "checker").Logger(),
+		state:     state,
+		execState: execState,
+		seals:     seals,
+	}
+}
+
+func (e *Engine) Ready() <-chan struct{} {
+	// make sure we will run into a crashloop if result gets inconsistent
+	// with sealed result.
+	err := e.checkLastSealed()
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("execution consistency check failed on startup")
+	}
+	return e.unit.Ready()
+}
+
+func (e *Engine) Done() <-chan struct{} {
+	return e.unit.Done()
+}
+
+// when a block is finalized check if the last sealed has been executed,
+// if it has been executed, check whether if the sealed result is consistent
+// with the executed result
+func (e *Engine) OnFinalizedBlock(block *model.Block) {
+	err := e.checkLastSealed()
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("execution consistency check failed")
+	}
+}
+
+func (e *Engine) checkLastSealed() error {
```

Comment:
> If a node is falling behind, then the last sealed block might not be executed yet. But in this case, the receipt won't be broadcasted, when the execution node eventually catch up, it will crash and the bad receipt won't be broadcasted.

I'm not convinced this is always true. We will detect the fork when we execute past our locally sealed block. But if we are executing behind our local finalization/sealing, we could fork and produce inconsistent receipts until we execute past our locally sealed block. 

> it will crash and the bad receipt won't be broadcasted.

It will crash and _a_ bad receipt won't be broadcast. But we may have already broadcast a previous bad receipt.

Reply:
I guess we at least have the guarantee that we won't broadcast a receipt that could possibly be included in a block, if it's behind sealing. That should be sufficient.

Reply:
Yeah, the node might have built a long fork until it learned the receipts are inconsistent with the sealed results, but at least these receipts will not be broadcasted. So it will crash before executing any unsealed block, and no bad receipts will be broadcasted.